lib/rex/proto/dcerpc/packet.rb

```diff
@@ -10,6 +10,110 @@ class Packet
 
   UUID = Rex::Proto::DCERPC::UUID
 
+  def self.make_auth(last_response, opts, ntlm_options, auth_client = Authentication.new)
+    ntlm3, auth_length = auth_client.auth_ntlm_3(last_response, opts, ntlm_options)
+
+    frag_length = 28+auth_length
+
+    # Create the alter_context request packet
+    buff =
+    [
+      5,      # major version 5
+      0,      # minor version 0
+      16,     # Auth3
+      3,      # flags
+      0x10000000,  # data representation
+      frag_length,     # frag length
+      auth_length,      # auth length
+      1,      # call id
+      0,  #padd
+    ].pack('CCCCNvvVV')
+
+    return buff+ntlm3, 0
+
+  end
+
+  def self.make_bind_auth(uuid, vers, domain, name, ntlm_options, auth_client = Authentication.new)
+    # Process the version strings ("1.0", 1.0, "1", 1)
+    bind_vers_maj, bind_vers_min = UUID.vers_to_nums(vers)
+    xfer_vers_maj, xfer_vers_min = UUID.vers_to_nums(UUID.xfer_syntax_vers)
+
+    if UUID.is? UUID.xfer_syntax_uuid
+      xfer_syntax_uuid = UUID.uuid_pack(UUID.xfer_syntax_uuid)
+    end
+
+    spnego, auth_length = auth_client.auth_ntlm_1(domain, name, ntlm_options)
+    frag_length = 80+auth_length
+
+    # Create the bind request packet
+    buff =
+    [
+      5,      # major version 5
+      0,      # minor version 0
+      11,     # bind type
+      3,      # flags
+      0x10000000,  # data representation
+      frag_length,     # frag length
+      auth_length,      # auth length
+      0,      # call id
+      5840,   # max xmit frag
+      5840,   # max recv frag
+      0,      # assoc group
+      1,      # num ctx items
+      0,      # context id
+      1,      # num trans items
+      UUID.uuid_pack(uuid),   # interface uuid
+      bind_vers_maj,       # interface major version
+      bind_vers_min,       # interface minor version
+      UUID.xfer_syntax_uuid,  # transfer syntax
+      xfer_vers_maj,       # syntax major version
+      xfer_vers_min,       # syntax minor version
+    ].pack('CCCCNvvVvvVVvvA16vvA16vv')
```

Comment:
This would be great to have `BinData` structures for these packets. `RubySMB` has great DCERPC packet implementations:
- https://github.com/rapid7/ruby_smb/blob/master/lib/ruby_smb/dcerpc/pdu_header.rb
- https://github.com/rapid7/ruby_smb/blob/master/lib/ruby_smb/dcerpc/bind.rb